src/main/java/org/haffson/adventofcode/days/day03/RectangleClaim.java

```diff
@@ -0,0 +1,83 @@
+package org.haffson.adventofcode.days.day03;
+
+import java.util.Objects;
+
+public class RectangleClaim {
+    private String id;
+    private int leftSpace;
+    private int topSpace;
+    private int width;
+    private int height;
+
+    public RectangleClaim(final String id, final int leftSpace, final int topSpace, final int width, final int height) {
+        this.id = id;
+        this.leftSpace = leftSpace;
+        this.topSpace = topSpace;
+        this.width = width;
+        this.height = height;
+    }
+
+    public String getId() {
+        return id;
+    }
+
+    public void setId(final String id) {
+        this.id = id;
+    }
+
+    public int getLeftSpace() {
+        return leftSpace;
+    }
+
+    public void setLeftSpace(final int leftSpace) {
+        this.leftSpace = leftSpace;
+    }
+
+    public int getTopSpace() {
+        return topSpace;
+    }
+
+    public void setTopSpace(final int topSpace) {
+        this.topSpace = topSpace;
+    }
+
+    public int getWidth() {
+        return width;
+    }
+
+    public void setWidth(final int width) {
+        this.width = width;
+    }
+
+    public int getHeight() {
+        return height;
+    }
+
+    public void setHeight(final int height) {
+        this.height = height;
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        final RectangleClaim that = (RectangleClaim) o;
+        return leftSpace == that.leftSpace && topSpace == that.topSpace && width == that.width && height == that.height && Objects.equals(id, that.id);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, leftSpace, topSpace, width, height);
+    }
+
+    @Override
+    public String toString() {
+        return "RectangleClaim{" +
+                "id='" + id + '\'' +
+                ", leftSpace=" + leftSpace +
+                ", topSpace=" + topSpace +
+                ", width=" + width +
+                ", height=" + height +
+                '}';
+    }
+}
```

Comment:
Mit diesem Datentyp hast du eine gute Option, um weitere Hilfsmethoden hinzuzufügen.

Beispiele:

- Wie groß ist die Fläche des Claims? => `public int area()`
- Überschneidung mit anderem Claim? => `public Optional<RectangleClaim> intersect(RectangleClaim otherClaim)`
  - Nur vorhanden, wenn es überhaupt eine Überschneidung gibt, daher `Optional`
  - Ergebnis ist wieder ein Claim (da die Überschneidung auch ein Rechteck ist); hier könnte man noch versuchen, Claim und Rectangle voneinander abzugrenzen und weitere Datentypen zu definieren
- Obere linke Ecke: `public Coordinate topLeft()`
  - `Coordinate` wäre ein neuer Typ und kapselt eine `x` und `y`-Koordinate
  - Könnte man dann analog für alle Ecken als Methode anbieten
- Enthaltene Punkte: `public List<Coordinate> getCoveredCoordinates()`
  - Wäre z.B. nützlich, wenn man Überschneidungen nur einmal zählen will; doppelt erfasste Koordinaten könnte man dann aussortieren (`distinct()`-Operation in Streams)

Mit den ganzen Hilfsmethoden sollte sich das Problem dann auch in einem Stream lösen lassen:

- Paare aller Claims erstellen (z.B. via extra Typ `Pair` und `flatMap`); ggf. weitere Hilfsmethode erstellen: das gleiche Problem gab es in der Aufgabe von gestern
- Überschneidung berechnen => ergibt ggf. ein neues Claim-Objekt
- Einträge ohne Überschneidungen verwerfen
- Covered Coordinates der Überschneidungen abfragen und via `flatMap` in den Stream überführen
- Doppelte Koordinaten entfernen: `distinct()`
- Übrige Koordinaten zählen: `count()`

Reply:
Sehr interessante Anmerkungen!
wie oben erwähnt, würde ich diese Ideen gerne für die nächsten Rätsel mitnehmen, da sonst das komplette Programm umgebaut werden müsste.
(Ich kann erstmal keine Rechtecke mehr sehen und brauche ein neues Rätsel :D)